spark-operator/src/main/java/org/apache/spark/kubernetes/operator/reconciler/SparkApplicationReconciler.java

```diff
@@ -0,0 +1,227 @@
+import io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource;
+import io.javaoperatorsdk.operator.processing.event.source.informer.Mappers;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.spark.kubernetes.operator.Constants;
+import org.apache.spark.kubernetes.operator.SparkApplication;
+import org.apache.spark.kubernetes.operator.controller.SparkApplicationContext;
+import org.apache.spark.kubernetes.operator.health.SentinelManager;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverReadyObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverRunningObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverStartObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverTimeoutObserver;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppCleanUpStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppInitStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppReconcileStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppResourceObserveStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppRunningStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppTerminatedStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppValidateStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.UnknownStateStep;
+import org.apache.spark.kubernetes.operator.utils.ApplicationStatusUtils;
+import org.apache.spark.kubernetes.operator.utils.LoggingUtils;
+import org.apache.spark.kubernetes.operator.utils.StatusRecorder;
+
+import static org.apache.spark.kubernetes.operator.reconciler.ReconcileProgress.completeAndDefaultRequeue;
+import static org.apache.spark.kubernetes.operator.reconciler.SparkReconcilerUtils.commonResourceLabelsStr;
+
+/**
+ * Reconciler for Spark Application.
+ * Performs sanity check on the app, identify the reconcile steps based on App status
+ * and execute the steps.
+ */
+@ControllerConfiguration
+@Slf4j
+@RequiredArgsConstructor
+public class SparkApplicationReconciler
+    implements Reconciler<SparkApplication>,
+    ErrorStatusHandler<SparkApplication>,
+    EventSourceInitializer<SparkApplication>,
+    Cleaner<SparkApplication> {
+  private final StatusRecorder statusRecorder;
+  private final SentinelManager<SparkApplication> sentinelManager;
+
+  @Override
+  public UpdateControl<SparkApplication> reconcile(SparkApplication sparkApplication,
+                                                   Context<SparkApplication> context)
+      throws Exception {
+    LoggingUtils.TrackedMDC trackedMDC = new LoggingUtils.TrackedMDC();
+    try {
+      trackedMDC.set(sparkApplication);
+      if (sentinelManager.handleSentinelResourceReconciliation(sparkApplication,
```

Comment:
We discussed briefly sentinels/canaries in  @gyfora. To my understanding at least one aspects that is handles is there no problem with informer watches (issues fabric8 had before ), those should not be the problem anymore, especially not with Kubernetes version >= 1.25. Such validations of caches we could add directly to Java Operator SDK. I would be also interested in other use cases of this features, and how much issues it catches after the fabric8 client informer fixes.